[CLANG]Add Scalable vectors for mfloat8_t

CarolineConcatto opened the pull request:
This patch adds these new vector sizes for sve:
    svmfloat8_t

According to the ARM ACLE PR#323[1].

[1] ARM-software/acle#323

CarolineConcatto committed:
[CLANG]Add Scalable vectors for mfloat8_t

This patch adds these new vector sizes for sve:
svmfloat8_t

According to the ARM ACLE PR#323[1].

[1] ARM-software/acle#323

CarolineConcatto committed:
Merge branch 'main' into sve_neon_fp8

paulwalker-arm commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -4339,7 +4339,6 @@ ASTContext::getBuiltinVectorTypeInfo(const BuiltinType *Ty) const {
   switch (Ty->getKind()) {
   default:
     llvm_unreachable("Unsupported builtin vector type");
-
```
Please can we keep these blank lines.  With all the continuation markers they help partition the related defines and includes.

paulwalker-arm commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -4354,12 +4353,16 @@ ASTContext::getBuiltinVectorTypeInfo(const BuiltinType *Ty) const {
     return {BoolTy, llvm::ElementCount::getScalable(NumEls), NF};
 #define SVE_OPAQUE_TYPE(Name, MangledName, Id, SingletonId)
 #include "clang/Basic/AArch64SVEACLETypes.def"
-
```
As above.

paulwalker-arm commented on clang/include/clang/Basic/AArch64SVEACLETypes.def:
```diff
@@ -69,6 +69,11 @@
+  SVE_VECTOR_TYPE_DETAILS(Name, MangledName, Id, SingletonId, NumEls, ElBits, NF, true, false, true)
+#endif
+
+#ifndef SVE_VECTOR_TYPE_MFLOAT
+#define SVE_VECTOR_TYPE_MFLOAT(Name, MangledName, Id, SingletonId, NumEls, ElBits, NF) \
```
Is this necessary.  Looking at the uses I think `SVE_VECTOR_TYPE_INT` would work?

paulwalker-arm commented on clang/utils/TableGen/SveEmitter.cpp:
```diff
@@ -87,6 +87,10 @@ class SVEType {
   bool isBFloat() const { return BFloat && !Float; }
+  bool isMFloat() const {
+    return MFloat && !BFloat && !Float;
+    ;
```
Stray `;`

paulwalker-arm commented on clang/utils/TableGen/SveEmitter.cpp:
```diff
@@ -87,6 +87,10 @@ class SVEType {
   bool isDefault() const { return DefaultType; }
   bool isFloat() const { return Float && !BFloat; }
   bool isBFloat() const { return BFloat && !Float; }
```
To be consistent I think these need updating to include `&& !MFloat`?

There's also an "isFloatingPoint()" method that probably needs `|| MFloat`.

paulwalker-arm commented on clang/utils/TableGen/SveEmitter.cpp:
```diff
@@ -454,6 +458,8 @@ std::string SVEType::builtin_str() const {
     assert(ElementBitwidth == 16 && "Not a valid BFloat.");
     S += "y";
+  } else if (isMFloat()) {
+    S += "m";
```
Perhaps worth adding `assert(ElementBitwidth == 8 && "Not a valid MFloat.")`?

CarolineConcatto replied:
Done

paulwalker-arm reviewed (commented):
Not sure if there's enough support to update all of them but, but many of the following should be extended to cover the new types:

clang/test/CodeGenCXX/aarch64-sve-vector-init.cpp
clang/test/CodeGenCXX/aarch64-mangle-sve-fixed-vectors.cpp
clang/test/CodeGenCXX/aarch64-mangle-sve-vectors.cpp
clang/test/CodeGenCXX/aarch64-sve-fixedtypeinfo.cpp
clang/test/CodeGenCXX/aarch64-sve-typeinfo.cpp
clang/test/AST/ast-dump-aarch64-sve-types.c
clang/test/CodeGen/aarch64-sve.c
clang/test/CodeGenObjC/aarch64-sve-types.m
clang/test/Sema/aarch64-sve-types.c
clang/test/Sema/attr-arm-sve-vector-bits.c
clang/test/SemaObjC/aarch64-sve-types.m

Is it worth adding the tuple types as well? I'm happy for that to be a separate PR but I figure you'll be touching the same files so it might be easier done together.  Your choice.

CarolineConcatto committed:
Address comments review

CarolineConcatto committed:
Merge branch 'main' into sve_neon_fp8

CarolineConcatto commented:
Hi @paulwalker-arm,
I believe I addressed all the comments in the code. 

I updated almost all of them to have a test for the SVMfloat8_t.
The ones that are not updated are the ones that test FIXED SVE types. 
The mangling needs scalar type support, BuiltinType::MFloat8,  in order to NOT mangle to SVUint8 (CXXNameMangler::mangleAArch64FixedSveVectorType(const VectorType *T))
I believe I can add support for that when the scalar type is merged.
About the tuple types.
If you dont mind I would rather address that in another patch.

paulwalker-arm commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -4419,6 +4419,13 @@ QualType ASTContext::getScalableVectorType(QualType EltTy, unsigned NumElts,
       EltTySize == ElBits && NumElts == (NumEls * NF) && NumFields == 1) {     \
     return SingletonId;                                                        \
   }
+#define SVE_VECTOR_TYPE_MFLOAT(Name, MangledName, Id, SingletonId, NumEls,     \
+                               ElBits, NF)                                     \
+  if (EltTy->hasIntegerRepresentation() && !EltTy->isBooleanType() &&          \
+      !EltTy->hasSignedIntegerRepresentation() && EltTySize == ElBits &&       \
+      NumElts == (NumEls * NF) && NumFields == 1) {                            \
+    return SingletonId;                                                        \
+  }
```
This is no longer required?

CarolineConcatto replied:
I missed that when I removed the MFLOAT.

paulwalker-arm commented on clang/include/clang/Basic/AArch64SVEACLETypes.def:
```diff
@@ -69,7 +69,7 @@
 #ifndef SVE_VECTOR_TYPE_BFLOAT
 #define SVE_VECTOR_TYPE_BFLOAT(Name, MangledName, Id, SingletonId, NumEls, ElBits, NF) \
-  SVE_VECTOR_TYPE_DETAILS(Name, MangledName, Id, SingletonId, NumEls, ElBits, NF, false, false, true)
+  SVE_VECTOR_TYPE_DETAILS(Name, MangledName, Id, SingletonId, NumEls, ElBits, NF, true, false, true)
```
Is this change necessary?

CarolineConcatto replied:
I missed that when I removed the MFLOAT.

CarolineConcatto committed:
Merge branch 'main' into sve_neon_fp8

CarolineConcatto committed:
Address review comments

CarolineConcatto committed:
Merge branch 'main' into sve_neon_fp8

CarolineConcatto committed:
Merge branch 'main' into sve_neon_fp8

CarolineConcatto committed:
Merge branch 'main' into sve_neon_fp8
